File: app/src/main/java/com/example/taxconnect/CommunityActivity.java
package com.example.taxconnect;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.taxconnect.adapter.PostAdapter;
import com.example.taxconnect.model.PostModel;
import com.example.taxconnect.model.UserModel;
import com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton;
import com.google.android.material.textfield.TextInputEditText;
import com.google.firebase.Timestamp;
import com.google.firebase.auth.FirebaseAuth;

import java.util.List;
import java.util.UUID;

public class CommunityActivity extends AppCompatActivity {

    private RecyclerView rvPosts;
    private PostAdapter adapter;
    private ExtendedFloatingActionButton fabCreatePost;
    private DataRepository repository;
    private UserModel currentUser;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_community);

        repository = DataRepository.getInstance();

        rvPosts = findViewById(R.id.rvPosts);
        fabCreatePost = findViewById(R.id.fabCreatePost);
        
        findViewById(R.id.toolbar).setOnClickListener(v -> finish());

        setupRecyclerView();
        fetchCurrentUser();
        // Posts will be loaded in onResume
        
        fabCreatePost.setOnClickListener(v -> showCreatePostDialog());
    }

    @Override
    protected void onResume() {
        super.onResume();
        loadPosts();
    }

    private void fetchCurrentUser() {
        String uid = FirebaseAuth.getInstance().getUid();
        if (uid == null) return;
        
        repository.fetchUser(uid, new DataRepository.DataCallback<UserModel>() {
            @Override
            public void onSuccess(UserModel data) {
                currentUser = data;
                if (adapter != null) {
                    adapter.setCurrentUserId(currentUser.getUid());
                }
            }

            @Override
            public void onError(String error) {
                // Handle error
            }
        });
    }

    private void setupRecyclerView() {
        adapter = new PostAdapter(new PostAdapter.OnPostClickListener() {
            @Override
            public void onPostClick(PostModel post) {
                openPostDetail(post);
            }

            @Override
            public void onLikeClick(PostModel post) {
                if (currentUser == null) return;
                
                repository.toggleLike(post.getId(), currentUser.getUid(), new DataRepository.DataCallback<Boolean>() {
                    @Override
                    public void onSuccess(Boolean isLiked) {
                        // Refresh specific item or whole list
                        // For simplicity, updating local model and notifying adapter
                        // But since others might like, reloading is safer but heavier.
                        // Let's update locally first for responsiveness.
                        List<String> likedBy = post.getLikedBy();
                        if (likedBy == null) likedBy = new java.util.ArrayList<>();
                        
                        if (isLiked) {
                            if (!likedBy.contains(currentUser.getUid())) likedBy.add(currentUser.getUid());
                        } else {
                            likedBy.remove(currentUser.getUid());
                        }
                        post.setLikedBy(likedBy);
                        post.setLikeCount(likedBy.size());
                        adapter.notifyDataSetChanged();
                    }

                    @Override
                    public void onError(String error) {
                        Toast.makeText(CommunityActivity.this, "Failed to update like: " + error, Toast.LENGTH_SHORT).show();
                    }
                });
            }

            @Override
            public void onCommentClick(PostModel post) {
                openPostDetail(post);
            }
        }, null);
        rvPosts.setLayoutManager(new LinearLayoutManager(this));
        rvPosts.setAdapter(adapter);
    }

    private void openPostDetail(PostModel post) {
        android.content.Intent intent = new android.content.Intent(this, PostDetailActivity.class);
        intent.putExtra("post", post);
        startActivity(intent);
    }

    private void loadPosts() {
        repository.getPosts(new DataRepository.DataCallback<List<PostModel>>() {
            @Override
            public void onSuccess(List<PostModel> data) {
                adapter.setPosts(data);
            }

            @Override
            public void onError(String error) {
                Toast.makeText(CommunityActivity.this, "Failed to load posts", Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void showCreatePostDialog() {
        if (currentUser == null) {
            Toast.makeText(this, "Please wait, loading user profile...", Toast.LENGTH_SHORT).show();
            return;
        }

        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        View view = LayoutInflater.from(this).inflate(R.layout.dialog_create_post, null);
        builder.setView(view);
        AlertDialog dialog = builder.create();

        // The ID in xml is etPostContent, correcting here
        TextInputEditText etPostContent = view.findViewById(R.id.etPostContent);
        
        Button btnPost = view.findViewById(R.id.btnPost);
        Button btnCancel = view.findViewById(R.id.btnCancel);

        btnPost.setOnClickListener(v -> {
            String content = etPostContent.getText().toString().trim();
            if (content.isEmpty()) return;

            createPost(content);
            dialog.dismiss();
        });

        btnCancel.setOnClickListener(v -> dialog.dismiss());
        dialog.show();
    }

    private void createPost(String content) {
        PostModel post = new PostModel();
        post.setId(UUID.randomUUID().toString());
        post.setUserId(currentUser.getUid());
        post.setUserName(currentUser.getName());
        post.setUserRole(currentUser.getRole());
        post.setContent(content);
        post.setTimestamp(new java.util.Date());
        post.setLikeCount(0);
        post.setCommentCount(0);

        repository.createPost(post, new DataRepository.DataCallback<Void>() {
            @Override
            public void onSuccess(Void data) {
                Toast.makeText(CommunityActivity.this, "Post Created!", Toast.LENGTH_SHORT).show();
                loadPosts();
            }

            @Override
            public void onError(String error) {
                Toast.makeText(CommunityActivity.this, "Failed to create post", Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: tax/app/src/main/java/com/example/taxconnect/features/community/CommunityActivity.kt
package com.example.taxconnect.features.community

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.Button
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.taxconnect.R
import com.example.taxconnect.databinding.ActivityCommunityBinding
import com.example.taxconnect.features.community.PostAdapter
import com.example.taxconnect.data.models.PostModel
import com.example.taxconnect.data.models.UserModel
import com.example.taxconnect.data.repositories.DataRepository
import com.example.taxconnect.core.base.BaseActivity
import com.google.android.material.textfield.TextInputEditText
import com.google.firebase.auth.FirebaseAuth
import java.util.Date
import java.util.UUID

class CommunityActivity : BaseActivity<ActivityCommunityBinding>() {

    override val bindingInflater: (LayoutInflater) -> ActivityCommunityBinding = ActivityCommunityBinding::inflate
    private lateinit var adapter: PostAdapter
    private var currentUser: UserModel? = null

    override fun initViews() {
        binding.toolbar.setNavigationOnClickListener { finish() }
        setupRecyclerView()
        fetchCurrentUser()
        // Posts will be loaded in onResume
    }

    override fun setupListeners() {
        binding.fabCreatePost.setOnClickListener { showCreatePostDialog() }
    }

    override fun observeViewModel() {
        // No ViewModel to observe yet
    }

    override fun onResume() {
        super.onResume()
        loadPosts()
    }

    private fun fetchCurrentUser() {
        val uid = FirebaseAuth.getInstance().uid ?: return

        DataRepository.getInstance().fetchUser(uid, object : DataRepository.DataCallback<UserModel> {
            override fun onSuccess(data: UserModel?) {
                currentUser = data
                if (data != null && ::adapter.isInitialized) {
                    adapter.setCurrentUserId(data.uid)
                }
            }

            override fun onError(error: String?) {
                // Handle error
            }
        })
    }

    private fun setupRecyclerView() {
        adapter = PostAdapter(object : PostAdapter.OnPostActionListener {
            override fun onPostClick(post: PostModel) {
                openPostDetail(post)
            }

            override fun onLikeClick(post: PostModel) {
                val user = currentUser ?: return
                val pid = post.id ?: return
                val uid = user.uid ?: return

                DataRepository.getInstance().toggleLike(pid, uid, object : DataRepository.DataCallback<Boolean> {
                    override fun onSuccess(isLiked: Boolean?) {
                        if (isLiked == null) return
                        
                        val likedBy = post.likedBy.toMutableList()
                        
                        if (isLiked) {
                            if (!likedBy.contains(uid)) likedBy.add(uid)
                        } else {
                            likedBy.remove(uid)
                        }
                        post.likedBy = likedBy
                        post.likeCount = likedBy.size
                        adapter.notifyDataSetChanged()
                    }

                    override fun onError(error: String?) {
                        showToast("Failed to update like: $error")
                    }
                })
            }

            override fun onCommentClick(post: PostModel) {
                openPostDetail(post)
            }
        })
        
        binding.rvPosts.layoutManager = LinearLayoutManager(this)
        binding.rvPosts.adapter = adapter
    }

    private fun openPostDetail(post: PostModel) {
        val intent = Intent(this, PostDetailActivity::class.java)
        intent.putExtra("post", post)
        startActivity(intent)
    }

    private fun loadPosts() {
        DataRepository.getInstance().getPosts(object : DataRepository.DataCallback<List<PostModel>> {
            override fun onSuccess(data: List<PostModel>?) {
                if (data != null) {
                    adapter.setPosts(data)
                }
            }

            override fun onError(error: String?) {
                showToast("Failed to load posts")
            }
        })
    }

    private fun showCreatePostDialog() {
        if (currentUser == null) {
            showToast(getString(R.string.loading_user_profile))
            return
        }

        val builder = AlertDialog.Builder(this)
        val view = LayoutInflater.from(this).inflate(R.layout.dialog_create_post, null)
        builder.setView(view)
        val dialog = builder.create()

        val etPostContent = view.findViewById<TextInputEditText>(R.id.etPostContent)
        val btnPost = view.findViewById<Button>(R.id.btnPost)
        val btnCancel = view.findViewById<Button>(R.id.btnCancel)

        btnPost.setOnClickListener {
            val content = etPostContent.text.toString().trim()
            if (content.isEmpty()) {
                etPostContent.error = getString(R.string.post_content_empty)
                etPostContent.requestFocus()
                return@setOnClickListener
            }

            createPost(content)
            dialog.dismiss()
        }

        btnCancel.setOnClickListener { dialog.dismiss() }
        dialog.show()
    }

    private fun createPost(content: String) {
        val user = currentUser ?: return
        
        val post = PostModel(
            id = UUID.randomUUID().toString(),
            userId = user.uid,
            userName = user.name,
            userRole = user.role,
            content = content,
            timestamp = Date(),
            likeCount = 0,
            commentCount = 0,
            likedBy = ArrayList()
        )

        DataRepository.getInstance().createPost(post, object : DataRepository.DataCallback<Void?> {
            override fun onSuccess(data: Void?) {
                showToast(getString(R.string.post_created))
                loadPosts()
            }

            override fun onError(error: String?) {
                showToast(getString(R.string.failed_to_create_post))
            }
        })
    }
}
